### PR TITLE
docs: use lib.extend instead of patching nixpkgs

### PR DESCRIPTION
**Pull request by Scrumplex**
Alternative to https://github.com/nix-community/nixvim/pull/1011

Importing a patched version of Nixpkgs causes two evaluations of Nixpkgs and IFD, which significantly slow down evaluations.
This uses `lib.extend`, which only recalculates the fix point of lib, instead of all of Nixpkgs and works just as well without IFD!

### CODE REVIEW

**Comment by Scrumplex** [comment]
With my changes:
```
Benchmark 1: env NIXPKGS_ALLOW_UNFREE=1 nix build --impure -vL .#docs --no-eval-cache
  Time (mean ± σ):      5.575 s ±  0.092 s    [User: 3.665 s, System: 0.923 s]
  Range (min … max):    5.468 s …  5.729 s    10 runs
```

```
Benchmark 1: env NIXPKGS_ALLOW_UNFREE=1 nix build --impure -vL .#docs --no-eval-cache
  Time (mean ± σ):      9.519 s ±  0.093 s    [User: 4.385 s, System: 1.534 s]
  Range (min … max):    9.398 s …  9.715 s    10 runs
```

File: docs/default.nix [hunk 1]
```diff
@@ -4,19 +4,37 @@
...
-      (pkgs.applyPatches {
-        name = "nixpkgs-nixvim-doc";
-        src = pkgs.path;
-        patches = [ ./either_recursive.patch ];
-      })
-      {
-        inherit (pkgs) system;
-        config.allowUnfree = true;
+
+  lib = pkgs.lib.extend (
```
Comment (MattSturgeon) [inline docs/default.nix:8]: I think a comment would be helpful here. Maybe something along the liens of:

```suggestion
  # Extend nixpkg's lib, so that we can handle recursive leaf types such as `either`
  lib = pkgs.lib.extend (
```

?

File: flake-modules/packages.nix [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
       rawModules,
       helpers,
       ...
     }:
     {
-      packages = import ../docs { inherit rawModules pkgs helpers; };
+      packages = import ../docs {
+        inherit rawModules helpers;
+        pkgs = pkgsUnfree;
+      };
```
Comment (MattSturgeon) [inline flake-modules/packages.nix:14]: This change seems unrelated. What's the motivation?

**Review by MattSturgeon — COMMENTED** [review]
Thanks, this is much better!

This should be easier to maintain, and the speed boost is the cherry on the cake!

It's also another step towards fixing #1154, though unfortunately there's still another IFD poking around somewhere, so this doesn't fix `nix flake show`.

I left some minor comments. I'd also like @traxys's review, as this was mostly implementation.

As for the [CI Failure](https://buildbot.nix-community.org/#/builders/191/builds/716), that looks spurious to me. Pushing again to re-trigger the tests may help.

**Comment by MattSturgeon** [comment]
Could you also add some of the justification to your commit message body? i.e. mention that the change:
- Speeds up building the docs
- Removes patching nixpkgs (improving maintainability)
- Removes an IFD (relates to #1154)

File: flake-modules/packages.nix (hunk 2 above)
Comment (traxys) [reply flake-modules/packages.nix:14]: we had in pkgsDoc the following: `        config.allowUnfree = true;`

File: docs/default.nix (hunk 1 above)
Comment (GaetanLepage) [reply docs/default.nix:8]: Yes, this comment would help.

File: flake-modules/packages.nix (hunk 2 above)
Comment (MattSturgeon) [reply flake-modules/packages.nix:14]: > we had in pkgsDoc the following: ` config.allowUnfree = true;`

Ah I missed that, thanks.

Do you know _why_ we need unfree? This is a good opportunity to add a comment explaining.

**Review by GaetanLepage — COMMENTED** [review]
LGTM. Good to merge once @MattSturgeon's requests will have been addressed.

File: flake-modules/packages.nix (hunk 2 above)
Comment (traxys) [reply flake-modules/packages.nix:14]: Because some plugins/dependencies are unfree I believe, and building the docs evaluates them

Comment (MattSturgeon) [reply flake-modules/packages.nix:14]: Maybe a comment along these lines then?
```nix
# Building the docs evaluates each plugin's default package, some of which are unfree
```

**Review by MattSturgeon — APPROVED** [review]
Thanks again!
